### PR TITLE
KAFKA-13032: add NPE checker for KeyValueMapper

### PR DESCRIPTION
**Pull request by jasonyanwenl**
Currently in both `KStreamMap` and `KStreamFlatMap` classes, they will throw NPE if the call to `KeyValueMapper#apply` return Null. We should check whether the result of that call is Null and throw a more meaningful error message for better debugging.

Two unit tests are also added to check if we successfully captured the Null.

@mjsax Please help review, thanks!

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamFlatMapTest.java
Comment (jlprat) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamFlatMapTest.java]: As the whole point of this PR is to provide better messages, I would also check in the test that the exception has the new enhanced message. Something like 
```suggestion
        final Record<String, Integer> record = new Record<>("K", 0, 0L);
        assertThrows(NullPointerException.class, () -> supplier.get().process(record), String.format("KeyValueMapper can't return null from mapping the record: %s", record));
```

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamMapTest.java
Comment (jlprat) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamMapTest.java]: Same as previous comment

**Review by jlprat — COMMENTED** [review]
Thanks for your PR @jasonyanwenl !
As the purpose of the PR is to provide better messages in case of null in these two particular places, I would say we should test for the message.

Otherwise the change looks good to me

**Review by showuon — APPROVED** [review]
LGTM! Thanks for the fix! 
One suggestion: please try not to email to the whole dev group to request for a code review next time (unless it's an urgent PR). If every PR owner sends an email in the dev group for code review, you can imagine how many emails will received for dev group members. Please try to mention the possible reviewer's names (could be more than one) in the PR, and try again next week ( or next 2 weeks, or more ) if no response. Hope that helps, and welcome to the Kafka community! :)

**Comment by jasonyanwenl** [comment]
> LGTM! Thanks for the fix!
> One suggestion: please try not to email to the whole dev group to request for a code review next time (unless it's an urgent PR). If every PR owner sends an email in the dev group for code review, you can imagine how many emails will received for dev group members. Please try to mention the possible reviewer's names (could be more than one) in the PR, and try again next week ( or next 2 weeks, or more ) if no response. Hope that helps, and welcome to the Kafka community! :)

Hi @showuon, thanks for your suggestion! Actualy when I checked the official site "[How to Contribute](https://kafka.apache.org/contributing)", I found it said "**Nag us if we aren't doing our job...**" and when I clicked the link "**Nag us**", it directly pops up an emailing window. So I guess we may need to update that webiste to avoid suggesting people directly sending email to the entire dev email group for PR review request (or be more clear that is for urgent fix).

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamFlatMapTest.java
Comment (jasonyanwenl) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamFlatMapTest.java]: @jlprat Thanks for your review! I found we mixed the using of two packages: `org.junit.Assert` and `org.junit.jupiter.api.Assertions`.  The method `asssertThrows` in those two packages have different api:

For `org.junit.Assert.assertThrows`, the [api](https://junit.org/junit4/javadoc/4.13/org/junit/Assert.html#assertThrows(java.lang.String,%20java.lang.Class,%20org.junit.function.ThrowingRunnable)) is:

```java
public static <T extends Throwable> T assertThrows(String message,
                                                   Class<T> expectedThrowable,
                                                   ThrowingRunnable runnable)
```

For `org.junit.jupiter.api.Assertions.assertThrows`, the [api](https://junit.org/junit5/docs/5.0.1/api/org/junit/jupiter/api/Assertions.html#assertThrows-java.lang.Class-org.junit.jupiter.api.function.Executable-java.lang.String-) is:
```java
public static <T extends Throwable> T assertThrows(Class<T> expectedType,
                                                   Executable executable,
                                                   String message)
```

I searched our Kafka code base and found there is [another place](https://github.com/apache/kafka/blob/9565a529e08d7aa36beac02c8e6115bcc87d2dc7/streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java#L521) using the package `org.junit.Assert.assertThrows` to assert exception message so I follow that to use the same api as well here.

Please let me know if this is not proper.

Comment (jlprat) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamFlatMapTest.java]: Actually I was wrong with my code suggestion, the overloaded method doesn't do what I expected it to do.
The message it takes is only to print if the assertion fails.

That should do the trick!
```
        final Throwable throwable = assertThrows(NullPointerException.class, () -> supplier.get().process(record));
        assertEquals(throwable.getMessage(), String.format("KeyValueMapper can't return null from mapping the record: %s", record));
```
Sorry about the confusion I brought with the first comment

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamFlatMapTest.java
Comment (jlprat) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamFlatMapTest.java]: Just copying over the suggestion to here, so it's easy to find

```suggestion
       final Throwable throwable = assertThrows(NullPointerException.class, () -> supplier.get().process(record));
       assertEquals(throwable.getMessage(), String.format("KeyValueMapper can't return null from mapping the record: %s", record));
```

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamFlatMapTest.java
Comment (jasonyanwenl) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamFlatMapTest.java]: @jlprat  I see. I didn't notice this, either LOL. Fixed!

**Review by jlprat — APPROVED** [review]
LGTM as long as both tests run green on Jenkins :)

Thanks for your PR @jasonyanwenl

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamFlatMap.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamFlatMap.java]: Just a suggestion:
```suggestion
            Objects.requireNonNull(newKeyValues, "The provided KeyValueMapper returned null which is not allowed.");
```
BTW, we should not output records since they might contain sensitive data.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamMap.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamMap.java]: Just a suggestion:
```suggestion
            Objects.requireNonNull(newPair, "The provided KeyValueMapper returned null which is not allowed.");
```

BTW, we should not output records since they might contain sensitive data.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamFlatMapTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamFlatMapTest.java]: We prefer to use `assertThat()`:

```suggestion
        assertThat(throwable.getMessage(), is(...);
```

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamMapTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamMapTest.java]: See my comment above.

**Review by cadonna — COMMENTED** [review]
Thanks for the PR @jasonyanwenl !

Here my feedback!

I think we should consider to introduce a specific exception for exceptions that originate from user code. Such exceptions could be distinguished from other exceptions when caught in the [Streams uncaught exception handler](https://kafka.apache.org/28/javadoc/org/apache/kafka/streams/errors/StreamsUncaughtExceptionHandler.html) and better reacted upon. 

Adding such an exception would require a KIP. Are you interested?

**Comment by jasonyanwenl** [comment]
> Thanks for the PR @jasonyanwenl !
> 
> Here my feedback!
> 
> I think we should consider to introduce a specific exception for exceptions that originate from user code. Such exceptions could be distinguished from other exceptions when caught in the [Streams uncaught exception handler](https://kafka.apache.org/28/javadoc/org/apache/kafka/streams/errors/StreamsUncaughtExceptionHandler.html) and better reacted upon.
> 
> Adding such an exception would require a KIP. Are you interested?

Hi @cadonna, thanks for yor feedback. Those are great insights! I'm definitely interested in doing this. But before that, have several questions:

1. Will this be the first case where we have a distiguished exception type for user code error?
2. Shall we finish this as a simple beginner issue and create a new ticket tracking the design & impl of that? The concern is that this NPE ticket might not be the only place where the new exceptioin could bring benefit and we probably should analyze code to expand the scope and update exception type all at once in a dedicated PR/ticket. Hence I think separating those two is suitable.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamFlatMapTest.java
Comment (jasonyanwenl) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamFlatMapTest.java]: 1. Just curious, why preferred `assertThat` from `hamcrest`? I saw in our code base, different places use `junit` / `hamcrest`. Any rule for guiding when to use which framework? 
2. Other testing in this file are using `junit`. So if I only use `hamcrest` in this single place, this class will import both `junit` and `hamcrest` and mix the using of those two. I think this is a bit ugly LOL. Do you suggest updating all to using `hamcrest`?

Thanks!

Comment (cadonna) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamFlatMapTest.java]: In Streams, we use `hamcrest` because I think most of us find `assertThat()` more readable. I think other components do not use `hamcrest`. This is just a minor comment and it will not block the merge. Anyways, if you want to use `assertEquals()` you need to swap the parameters, since with `assertEquals()` the first parameter is the expected value and not the actual value.

**Comment by cadonna** [comment]
> > Thanks for the PR @jasonyanwenl !
> > Here my feedback!
> > I think we should consider to introduce a specific exception for exceptions that originate from user code. Such exceptions could be distinguished from other exceptions when caught in the [Streams uncaught exception handler](https://kafka.apache.org/28/javadoc/org/apache/kafka/streams/errors/StreamsUncaughtExceptionHandler.html) and better reacted upon.
> > Adding such an exception would require a KIP. Are you interested?
> 
> Hi @cadonna, thanks for yor feedback. Those are great insights! I'm definitely interested in doing this. But before that, have several questions:
> 
>     1. Will this be the first case where we have a distiguished exception type for user code error?
> 
>     2. Shall we finish this as a simple beginner issue and create a new ticket tracking the design & impl of that? The concern is that this NPE ticket might not be the only place where the new exceptioin could bring benefit and we probably should analyze code to expand the scope and update exception type all at once in a dedicated PR/ticket. Hence I think separating those two is suitable.

1. Yes
2. I agree that you should first finish this PR/ticket and then work on a more holistic approach for user code exceptions.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamFlatMapTest.java
Comment (jasonyanwenl) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamFlatMapTest.java]: Got it. Thanks for your explanation. I updated to use the assertThat.

**Comment by jasonyanwenl** [comment]
Hi @cadonna Could you please help review/approve/merge this PR? Thanks!

As for the user exception task we discussed before, I was trying to create a ticket but I realized that I could not fully understand the requirement and ideal outcome of that task. For exp, I'm not clear what kind of action we should do if encoutering such exception. 

Anyway, could I just create a ticket with a rough description (or maybe you could help me create one if you get a chance?) and let's start a thread discussion in that ticket?

Thanks again!

**Review by cadonna — APPROVED** [review]
LGTM! Thank you, @jasonyanwenl !

I re-triggered the builds since there was a build that exited unexpectedly.

If the the build results are acceptable, I will merge this PR.

**Comment by cadonna** [comment]
> As for the user exception task we discussed before, I was trying to create a ticket but I realized that I could not fully understand the requirement and ideal outcome of that task. For exp, I'm not clear what kind of action we should do if encoutering such exception.

The task basically consists of introducing a public exception that is used to identify exceptions originating from user code in Streams. This exception should be thrown if user code throws an exception and contain the original exception as its cause. Users decide how to react on this exception in their custom uncaught exception handler. Since a public exception is a change of the public API, we need a KIP to introduce it.

> Anyway, could I just create a ticket with a rough description (or maybe you could help me create one if you get a chance?) and let's start a thread discussion in that ticket?

That is definitely a good start.

**Pull request merged**
